### PR TITLE
[TACHYON-937] Add source location when moving or removing a block

### PR DESCRIPTION
**Pull request by shimingfei**
https://tachyon.atlassian.net/browse/TACHYON-937

### CODE REVIEW

**Comment by calvinjia** [comment]
@shimingfei Thanks for the patch, could you take a look at the test failures?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java [hunk 1]
```diff
@@ -27,20 +27,21 @@
  */
 public final class EvictionPlan {
   /** A list of pairs of blockId and its location to move to */
-  private final List<Pair<Long, BlockStoreLocation>> mToMove;
+  private final List<Pair<Long, Pair<BlockStoreLocation, BlockStoreLocation>>> mToMove;
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java]: I think the type is complex enough to define an inner class for it.

File: servers/src/main/java/tachyon/worker/block/TieredBlockStore.java [hunk 2]
```diff
@@ -796,6 +820,11 @@ private void removeBlockInternal(long userId, long blockId) throws InvalidStateE
         mMetadataReadLock.unlock();
       }
 
+      if (!location.equals(blockMeta.getBlockLocation())
+          && !location.equals(BlockStoreLocation.anyTier())) {
+        LOG.info("Failed to remove block, because the location is changed!");
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: It might be better to throw an exception than logging the failure.

Comment (shimingfei) [inline servers/src/main/java/tachyon/worker/block/TieredBlockStore.java]: yes, an exception is better. Thanks!

File: servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java (hunk 1 above)
Comment (shimingfei) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java]: I will add new class for it.

File: servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java [hunk 3]
```diff
@@ -27,26 +27,28 @@
...
-  public EvictionPlan(List<Pair<Long, BlockStoreLocation>> toTransfer, List<Long> toEvict) {
+  public EvictionPlan(List<BlockTransferInfo> toTransfer,
+      List<Pair<Long, BlockStoreLocation>> toEvict) {
     mToMove = Preconditions.checkNotNull(toTransfer);
     mToEvict = Preconditions.checkNotNull(toEvict);
   }
 
   /**
-   * @return a list of pairs of a block id and the location to move it to
+   * @return a list of pairs of a block id and the locations of where it comes from and will be
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java]: Could you update the doc accordingly?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java [hunk 4]
```diff
@@ -27,26 +27,28 @@
  */
 public final class EvictionPlan {
   /** A list of pairs of blockId and its location to move to */
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java]: Could you update the doc accordingly?

File: servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java [hunk 5]
```diff
@@ -27,26 +27,28 @@
...
-  private final List<Pair<Long, BlockStoreLocation>> mToMove;
+  private final List<BlockTransferInfo> mToMove;
   /** A list of blockId to remove */
```
Comment (yupeng9) [inline servers/src/main/java/tachyon/worker/block/evictor/EvictionPlan.java]: Could you update the doc accordingly?

**Comment by yupeng9** [comment]
Thanks for patch! I have a few minor comments left.

**Comment by shimingfei** [comment]
@yupeng9  the code is updated. Thanks!

**Comment by yupeng9** [comment]
Thanks. In the future, please keep the git history (e.g. avoid amending changes), so that we can better track the comments on GitHub.

**Comment by yupeng9** [comment]
+1 LGTM

**Comment by yupeng9** [comment]
@shimingfei, could you resolve the conflicts? The PR looks good to me.

**Comment by shimingfei** [comment]
@yupeng9 The PR is updated. Thanks!

**Comment by yupeng9** [comment]
+1 LGTM

**Pull request merged**